src/rust/bitbox02-rust-c/src/util.rs

```diff
@@ -119,46 +115,36 @@ impl AsRef<str> for CStrMut {
     /// with 0 length due to limitation in `core::slice`.
     fn as_ref(&self) -> &str {
         let mut buf = self.buf;
-        if self.len == 0 && self.buf.is_null() {
+        if self.cap == 0 && self.buf.is_null() {
             buf = core::ptr::NonNull::dangling().as_ptr();
         }
         assert!(!buf.is_null());
 
-        // Find null terminator
-        let mut len = 0;
-        unsafe {
-            let mut b = buf;
-            while b.read() != 0 {
-                b = b.offset(1);
-                len += 1;
-                if len == self.len {
-                    panic!("CStrMut not null terminated");
-                }
-            }
-        }
-
-        unsafe { core::str::from_utf8_unchecked(core::slice::from_raw_parts(buf, len)) }
+        unsafe { core::str::from_utf8_unchecked(core::slice::from_raw_parts(self.buf, self.len)) }
     }
 }
 
 impl CStrMut {
-    /// Create a slice to a buffer. Only allowed for non-null pointers with length or null pointers
-    /// with 0 length due to limitation in `core::slice`.
-    /// The caller must ensure that the string is null terminated.
-    pub unsafe fn as_bytes_mut(&mut self) -> &mut [u8] {
-        let mut buf = self.buf;
-        if self.len == 0 && self.buf.is_null() {
-            buf = core::ptr::NonNull::dangling().as_ptr();
+    /// Create a mutable slice to an unused range of the buffer. Will ensure that byte immediately
+    /// following the range is null.
+    pub fn unused(&mut self, req: usize) -> &mut [u8] {
+        // Must be room for requested amount of bytes and null terminator.
+        if self.cap - self.len < req + 1 {
+            panic!("Not enough bytes left in buffer");
+        }
+        let len = self.len;
+        self.len += req;
+        unsafe {
+            self.buf.offset((len + req) as isize).write(0);
+            core::slice::from_raw_parts_mut(self.buf.offset(len as isize), req)
```

Comment:
should be outside of `unsafe{}` so the caller has to mark the section unsafe, as they have to promise not to insert a null terminator, and actually to promise to fill the bytes (they might be all null to start with). otherwise `self.len` is wrong.

Could an alternative be to not return a `&mut [u8]`, but to do sth like this?

`cstrmut.unused(|&bytes| hex::encode_to_slice(&buf, &bytes))`? Then we could run the callback, and afterwards check that there is no null in `bytes`, making it safe.

Reply:
You are right, the user could write null into the byte array.. so currently this function would have to be marked unsafe.

We could also have a "write(&[u8])" function that does all the checks. A bit like "FixedCString" I've already implemented.